source/common/upstream/subset_lb.h

```diff
@@ -0,0 +1,117 @@
+#pragma once
+
+#include <functional>
+#include <map>
+#include <memory>
+#include <string>
+
+#include "envoy/common/optional.h"
+#include "envoy/runtime/runtime.h"
+#include "envoy/upstream/load_balancer.h"
+
+#include "common/common/macros.h"
+#include "common/protobuf/protobuf.h"
+#include "common/protobuf/utility.h"
+#include "common/upstream/upstream_impl.h"
+
+namespace Envoy {
+namespace Upstream {
+
+class SubsetLoadBalancer : public LoadBalancer, Logger::Loggable<Logger::Id::upstream> {
+public:
+  SubsetLoadBalancer(LoadBalancerType lb_type, HostSet& host_set, const HostSet* local_host_set,
+                     ClusterStats& stats, Runtime::Loader& runtime,
+                     Runtime::RandomGenerator& random, const LoadBalancerSubsetInfo& subsets);
+
+  // Upstream::LoadBalancer
+  HostConstSharedPtr chooseHost(LoadBalancerContext* context) override;
+
+private:
+  typedef std::function<bool(const Host&)> HostPredicate;
+
+  // Represents a subset of an original HostSet.
+  class HostSubsetImpl : public HostSetImpl {
+  public:
+    HostSubsetImpl(const HostSet& original_host_set)
+        : HostSetImpl(), original_host_set_(original_host_set), empty_(true) {}
+
+    void update(const std::vector<HostSharedPtr>& hosts_added,
+                const std::vector<HostSharedPtr>& hosts_removed, HostPredicate predicate);
+
+    void triggerCallbacks() { HostSetImpl::runUpdateCallbacks({}, {}); }
+
+    bool empty() { return empty_; };
+
+  private:
+    const HostSet& original_host_set_;
+    bool empty_;
```

Comment:
Do we need this var? Can we just look inside HostSetImpl hosts() to see if it's empty?